File: android/app/src/main/java/com/familyrecipe/core/repositories/RecipeRepository.kt
package com.familyrecipe.core.repositories

import com.familyrecipe.core.database.RecipeDao
import com.familyrecipe.core.models.Recipe
import com.familyrecipe.core.models.RecipeCategory
import com.familyrecipe.core.models.SyncStatus
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map
import java.util.Date

// MARK: - Recipe Repository

class RecipeRepository(
    private val recipeDao: RecipeDao
) {
    // MARK: - Read Operations

    fun getRecipes(familyId: String): Flow<List<Recipe>> =
        recipeDao.getRecipesByFamily(familyId)

    suspend fun getRecipesOnce(familyId: String): List<Recipe> =
        recipeDao.getRecipesByFamilyOnce(familyId)

    fun getRecipe(id: String): Flow<Recipe?> =
        recipeDao.getRecipeById(id)

    suspend fun getRecipeOnce(id: String): Recipe? =
        recipeDao.getRecipeByIdOnce(id)

    fun getRecipesByCategory(familyId: String, category: RecipeCategory): Flow<List<Recipe>> =
        recipeDao.getRecipesByCategory(familyId, category)

    fun getRecentRecipes(familyId: String, limit: Int = 10): Flow<List<Recipe>> =
        recipeDao.getRecentRecipes(familyId, limit)

    fun getFavoriteRecipes(familyId: String, memberId: String): Flow<List<Recipe>> =
        recipeDao.getFavoriteRecipes(familyId, memberId)

    fun searchRecipes(familyId: String, query: String): Flow<List<Recipe>> =
        recipeDao.searchRecipes(familyId, query)

    fun getRecipesByCreator(memberId: String): Flow<List<Recipe>> =
        recipeDao.getRecipesByCreator(memberId)

    suspend fun getRecipeCount(familyId: String): Int =
        recipeDao.getRecipeCount(familyId)

    // MARK: - Write Operations

    suspend fun createRecipe(recipe: Recipe) {
        recipeDao.insert(recipe)
    }

    suspend fun updateRecipe(recipe: Recipe) {
        recipeDao.update(recipe.copy(updatedAt = Date()))
    }

    suspend fun deleteRecipe(recipe: Recipe) {
        recipeDao.delete(recipe)
    }

    suspend fun deleteRecipeById(id: String) {
        recipeDao.deleteById(id)
    }

    // MARK: - Favorites

    suspend fun toggleFavorite(recipeId: String, memberId: String) {
        val recipe = recipeDao.getRecipeByIdOnce(recipeId) ?: return
        val updatedFavorites = if (recipe.favoritedBy.contains(memberId)) {
            recipe.favoritedBy.filter { it != memberId }
        } else {
            recipe.favoritedBy + memberId
        }
        recipeDao.update(recipe.copy(
            favoritedBy = updatedFavorites,
            updatedAt = Date()
        ))
    }

    suspend fun isFavorite(recipeId: String, memberId: String): Boolean {
        val recipe = recipeDao.getRecipeByIdOnce(recipeId) ?: return false
        return recipe.favoritedBy.contains(memberId)
    }

    // MARK: - Sync

    suspend fun getUnsyncedRecipes(): List<Recipe> =
        recipeDao.getUnsyncedRecipes()

    suspend fun updateSyncStatus(id: String, status: SyncStatus) {
        recipeDao.updateSyncStatus(id, status)
    }

    suspend fun markAsSynced(ids: List<String>) {
        recipeDao.markAsSynced(ids)
    }

    // MARK: - Bulk Operations

    suspend fun insertAll(recipes: List<Recipe>) {
        recipeDao.insertAll(recipes)
    }

    suspend fun deleteAllByFamily(familyId: String) {
        recipeDao.deleteAllByFamily(familyId)
    }

    // MARK: - Statistics

    suspend fun getRecipeCountByCategory(familyId: String): Map<RecipeCategory, Int> {
        val recipes = getRecipesOnce(familyId)
        return RecipeCategory.values().associateWith { category ->
            recipes.count { it.category == category }
        }
    }
}

File: ios/Core/Repositories/RecipeRepository.swift
import Foundation
import SwiftData

// MARK: - Recipe Repository Protocol

protocol RecipeRepositoryProtocol {
    func getRecipes(familyId: UUID) async throws -> [Recipe]
    func getRecipe(id: UUID) async throws -> Recipe?
    func getRecipesByCategory(familyId: UUID, category: RecipeCategory) async throws -> [Recipe]
    func getRecentRecipes(familyId: UUID, limit: Int) async throws -> [Recipe]
    func getFavoriteRecipes(familyId: UUID, memberId: UUID) async throws -> [Recipe]
    func searchRecipes(familyId: UUID, query: String) async throws -> [Recipe]
    func createRecipe(_ recipe: Recipe) async throws
    func updateRecipe(_ recipe: Recipe) async throws
    func deleteRecipe(_ recipe: Recipe) async throws
    func toggleFavorite(recipeId: UUID, memberId: UUID) async throws
}

// MARK: - Recipe Repository Implementation

@MainActor
final class RecipeRepository: RecipeRepositoryProtocol {
    private let modelContext: ModelContext

    init(modelContext: ModelContext) {
        self.modelContext = modelContext
    }

    func getRecipes(familyId: UUID) async throws -> [Recipe] {
        let descriptor = FetchDescriptor<Recipe>(
            predicate: #Predicate { $0.familyId == familyId },
            sortBy: [SortDescriptor(\.updatedAt, order: .reverse)]
        )
        return try modelContext.fetch(descriptor)
    }

    func getRecipe(id: UUID) async throws -> Recipe? {
        let descriptor = FetchDescriptor<Recipe>(
            predicate: #Predicate { $0.id == id }
        )
        return try modelContext.fetch(descriptor).first
    }

    func getRecipesByCategory(familyId: UUID, category: RecipeCategory) async throws -> [Recipe] {
        let descriptor = FetchDescriptor<Recipe>(
            predicate: #Predicate { $0.familyId == familyId && $0.category == category },
            sortBy: [SortDescriptor(\.title)]
        )
        return try modelContext.fetch(descriptor)
    }

    func getRecentRecipes(familyId: UUID, limit: Int) async throws -> [Recipe] {
        var descriptor = FetchDescriptor<Recipe>(
            predicate: #Predicate { $0.familyId == familyId },
            sortBy: [SortDescriptor(\.createdAt, order: .reverse)]
        )
        descriptor.fetchLimit = limit
        return try modelContext.fetch(descriptor)
    }

    func getFavoriteRecipes(familyId: UUID, memberId: UUID) async throws -> [Recipe] {
        let allRecipes = try await getRecipes(familyId: familyId)
        return allRecipes.filter { $0.favoritedBy.contains(memberId) }
    }

    func searchRecipes(familyId: UUID, query: String) async throws -> [Recipe] {
        let lowercasedQuery = query.lowercased()
        let allRecipes = try await getRecipes(familyId: familyId)

        return allRecipes.filter { recipe in
            recipe.title.lowercased().contains(lowercasedQuery) ||
            recipe.recipeDescription.lowercased().contains(lowercasedQuery) ||
            recipe.tags.contains { $0.lowercased().contains(lowercasedQuery) }
        }
    }

    func createRecipe(_ recipe: Recipe) async throws {
        modelContext.insert(recipe)
        try modelContext.save()
    }

    func updateRecipe(_ recipe: Recipe) async throws {
        recipe.updatedAt = Date()
        try modelContext.save()
    }

    func deleteRecipe(_ recipe: Recipe) async throws {
        modelContext.delete(recipe)
        try modelContext.save()
    }

    func toggleFavorite(recipeId: UUID, memberId: UUID) async throws {
        guard let recipe = try await getRecipe(id: recipeId) else { return }

        if recipe.favoritedBy.contains(memberId) {
            recipe.favoritedBy.removeAll { $0 == memberId }
        } else {
            recipe.favoritedBy.append(memberId)
        }

        recipe.updatedAt = Date()
        try modelContext.save()
    }
}

// MARK: - Recipe Count by Category

extension RecipeRepository {
    func getRecipeCountByCategory(familyId: UUID) async throws -> [RecipeCategory: Int] {
        let allRecipes = try await getRecipes(familyId: familyId)
        var counts: [RecipeCategory: Int] = [:]

        for category in RecipeCategory.allCases {
            counts[category] = allRecipes.filter { $0.category == category }.count
        }

        return counts
    }

    func getTotalRecipeCount(familyId: UUID) async throws -> Int {
        try await getRecipes(familyId: familyId).count
    }
}
